File: scripts/hw6_511.R
# Load required libraries
library(rvest)
library(dplyr)
library(ggplot2)
library(reshape2)

# Step 1: Fetch the webpage content
url <- 'https://www.basketball-reference.com/leagues/NBA_2024_per_game.html'
webpage <- read_html(url)

# Step 2: Parse the HTML content and extract the table
table <- webpage %>% html_node("#per_game_stats") %>% html_table(fill = TRUE)

# Step 3: Convert to a DataFrame
df <- as.data.frame(table)

# Step 4: Save the dataset locally for reproducibility
write.csv(df, "NBA_2024_per_game.csv", row.names = FALSE)
cat("Dataset saved as NBA_2024_per_game.csv\n")

# Step 5: Clean the DataFrame
df <- df %>%
  filter(Player != "" & Player != "Player") %>%
  mutate(across(where(is.character), as.character)) %>%
  mutate(across(where(is.numeric), ~ replace_na(., 0)))

# Step 6: Convert relevant columns to numeric for analysis
numeric_columns <- c('PTS', 'AST', 'TOV', 'TRB', 'STL', 'BLK')
df <- df %>%
  mutate(across(all_of(numeric_columns), as.numeric))

# Step 7: Check column names and availability for plotting
cat("Available Columns:", colnames(df), "\n")
missing_columns <- setdiff(numeric_columns, colnames(df))
if (length(missing_columns) > 0) {
  cat("Some columns are missing:", missing_columns, "\n")
}

# Step 8: Plot 1 - Points per Game Distribution
ggplot(df, aes(x = PTS)) +
  geom_histogram(bins = 20, fill = "blue", color = "black", alpha = 0.7) +
  labs(title = "Distribution of Points per Game", x = "Points per Game", y = "Frequency") +
  theme_minimal()

# Step 9: Plot 2 - Assists vs. Turnovers
ggplot(df, aes(x = AST, y = TOV, color = Pos)) +
  geom_point(alpha = 0.7) +
  labs(title = "Assists vs. Turnovers by Position", x = "Assists per Game", y = "Turnovers per Game") +
  theme_minimal()

# Step 10: Plot 3 - Correlation Heatmap
correlation_data <- df %>% select(all_of(numeric_columns))
correlation_matrix <- cor(correlation_data, use = "complete.obs")
cor_melt <- melt(correlation_matrix, varnames = c("Var1", "Var2"), value.name = "value")

ggplot(cor_melt, aes(x = Var1, y = Var2, fill = value)) +
  geom_tile(color = "gray80", size = 0.5) +
  geom_text(aes(label = sprintf("%.2f", value)), color = "black", size = 3) +
  scale_fill_gradient2(
    low = "blue", high = "red", mid = "white",
    midpoint = 0, limit = c(-1, 1), space = "Lab",
    name = "Correlation"
  ) +
  labs(title = "Correlation Heatmap of Selected NBA Stats (2024)", x = "Statistics", y = "Statistics") +
  theme_minimal() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1, size = 10),
    axis.text.y = element_text(size = 10),
    panel.grid = element_blank(),
    legend.position = "right"
  )

# Step 11: Plot 4 - Top Scorers (Bar Plot)
top_scorers <- df %>% top_n(10, PTS)
ggplot(top_scorers, aes(x = reorder(Player, PTS), y = PTS, fill = Player)) +
  geom_bar(stat = "identity", show.legend = FALSE) +
  coord_flip() +
  labs(title = "Top 10 Scorers in the NBA (Per Game)", x = "Player", y = "Points Per Game") +
  theme_minimal()

# Step 12: Radar Chart - Skill Comparison
radar_chart <- function(data, categories, player_name) {
  player_data <- data %>% filter(Player == player_name) %>% select(all_of(categories))
  if (nrow(player_data) == 0) {
    cat("Player", player_name, "not found in the dataset.\n")
    return()
  }
  player_values <- as.numeric(unlist(player_data))
  player_values <- c(player_values, player_values[1])  # Close the radar chart
  
  angles <- seq(0, 2 * pi, length.out = length(player_values))
  radar_df <- data.frame(
    category = c(categories, categories[1]),
    value = player_values,
    angle = angles
  )
  
  ggplot(radar_df, aes(x = angle, y = value)) +
    geom_polygon(fill = "blue", alpha = 0.3) +
    geom_line(color = "blue") +
    geom_point(color = "red") +
    coord_polar(start = 0) +
    scale_x_continuous(breaks = angles[-length(angles)], labels = categories) +
    labs(title = paste(player_name, "Skill Comparison")) +
    theme_minimal()
}

# Define categories and generate radar chart for LeBron James
radar_chart(df, numeric_columns, "LeBron James")

# Step 13: Plot 5 - Points vs. Rebounds by Position
ggplot(df, aes(x = PTS, y = TRB, size = AST, color = Pos)) +
  geom_point(alpha = 0.7) +
  scale_size_continuous(range = c(1, 10)) +
  labs(title = "Points vs. Rebounds (Bubble Size = Assists)", x = "Points Per Game", y = "Rebounds Per Game") +
  theme_minimal()

# Step 14: Plot 6 - Box Plot (Points by Position)
ggplot(df, aes(x = Pos, y = PTS, fill = Pos)) +
  geom_boxplot() +
  labs(title = "Points Per Game by Position", x = "Position", y = "Points Per Game") +
  theme_minimal()

File: hw6_511.py
import requests
import pandas as pd
from bs4 import BeautifulSoup
import matplotlib.pyplot as plt
import seaborn as sns
from math import pi

# Step 1: Fetch the webpage content
url = 'https://www.basketball-reference.com/leagues/NBA_2024_per_game.html'
response = requests.get(url)
response.raise_for_status()  # Ensure the request was successful

# Step 2: Parse the HTML content using BeautifulSoup
soup = BeautifulSoup(response.text, 'html.parser')
table = soup.find('table', {'id': 'per_game_stats'})

# Step 3: Read the table into a Pandas DataFrame
df = pd.read_html(str(table))[0]

# Step 4: Clean the DataFrame
df = df.dropna(subset=['Player'])  # Remove rows where 'Player' is NaN
df = df[df['Player'] != 'Player']  # Remove header rows that are repeated within the table
df = df.fillna(0)  # Fill NaN values with 0
df = df.apply(pd.to_numeric, errors='ignore')  # Convert numeric columns to appropriate data types

# Step 5: Print column names to verify
print("Available Columns:", df.columns)

# Step 6: Corrected column selection based on actual DataFrame
columns_to_plot = ['PTS', 'AST', 'TOV', 'TRB', 'STL', 'BLK']  # Replace 'REB' with 'TRB'

# Check if all required columns exist
existing_columns = [col for col in columns_to_plot if col in df.columns]
if len(existing_columns) < len(columns_to_plot):
    print(f"Some columns are missing: {[col for col in columns_to_plot if col not in df.columns]}")

# Step 7: Set up plotting aesthetics
sns.set(style='whitegrid')

# Plot 1: Points per Game Distribution
plt.figure(figsize=(12, 8))
sns.histplot(df['PTS'], bins=20, kde=True, color='blue')
plt.title('Distribution of Points per Game')
plt.xlabel('Points per Game')
plt.ylabel('Frequency')
plt.show()

# Plot 2: Assists vs. Turnovers (Scatter Plot)
plt.figure(figsize=(12, 8))
sns.scatterplot(data=df, x='AST', y='TOV', hue='Pos', palette='deep')
plt.title('Assists vs. Turnovers by Position')
plt.xlabel('Assists per Game')
plt.ylabel('Turnovers per Game')
plt.legend(title='Position')
plt.show()

# Plot 3: Correlation Heatmap
plt.figure(figsize=(12, 10))
correlation_matrix = df[existing_columns].corr()
sns.heatmap(correlation_matrix, annot=True, cmap='coolwarm', fmt='.2f')
plt.title('Correlation Heatmap of Key Stats')
plt.show()

# Plot 4: Top Scorers (Bar Plot)
top_scorers = df.nlargest(10, 'PTS')
plt.figure(figsize=(14, 8))
sns.barplot(data=top_scorers, x='PTS', y='Player', palette='viridis')
plt.title('Top 10 Scorers in the NBA (Per Game)')
plt.xlabel('Points Per Game')
plt.ylabel('Player')
plt.show()

# Plot 5: Radar Chart (Skill Comparison)
def radar_chart(data, categories, player_name):
    values = data.loc[data['Player'] == player_name, categories].values.flatten().tolist()
    if not values:
        print(f"Player {player_name} not found in the dataset.")
        return
    values += values[:1]  # Repeat the first value for a closed polygon
    
    angles = [n / float(len(categories)) * 2 * pi for n in range(len(categories))]
    angles += angles[:1]
    
    plt.figure(figsize=(6, 6))
    ax = plt.subplot(111, polar=True)
    plt.xticks(angles[:-1], categories, color='grey', size=8)
    ax.plot(angles, values, linewidth=2, linestyle='solid', label=player_name)
    ax.fill(angles, values, alpha=0.4)
    plt.title(f'{player_name} Skill Comparison')
    plt.legend(loc='upper right', bbox_to_anchor=(0.1, 0.1))
    plt.show()

# Radar Chart for a Specific Player
categories = ['PTS', 'AST', 'TRB', 'STL', 'BLK']
radar_chart(df, categories, 'LeBron James')  # Replace with any player name in the dataset

# Plot 6: Points vs. Rebounds by Position (Bubble Plot)
plt.figure(figsize=(12, 8))
sns.scatterplot(data=df, x='PTS', y='TRB', size='AST', hue='Pos', sizes=(20, 200), alpha=0.7)
plt.title('Points vs. Rebounds (Bubble Size = Assists)')
plt.xlabel('Points Per Game')
plt.ylabel('Rebounds Per Game')
plt.legend(title='Position', bbox_to_anchor=(1.05, 1), loc='upper left')
plt.show()

# Plot 7: Box Plot (Points by Position)
plt.figure(figsize=(12, 8))
sns.boxplot(data=df, x='Pos', y='PTS', palette='Set2')
plt.title('Points Per Game by Position')
plt.xlabel('Position')
plt.ylabel('Points Per Game')
plt.show()



